connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java

```diff
@@ -131,9 +133,26 @@ public void run() {
         }
     }
 
+    /**
+     * Fail the connector.
+     * @param cause the cause of the failure; if null, the connector will not be failed
+     */
+    public void fail(Throwable cause) {
+        synchronized (this) {
+            if (this.externalFailure != null)
+                return;
+            this.externalFailure = cause;
+            notify();
+        }
+    }
+
     void doRun() {
         initialize();
         while (!stopping) {
+            Throwable failure = externalFailure;
+            if (failure != null)
+                onFailure(failure);
```

Comment:
Is this in danger of being called more than once, particularly if the connector has this problem and then a pause/resume request comes in? Is that a bad thing?

It looks like the connector thread just waits in this loop until something external calls shutdown(), so I would expect this to get called whenever someone notify()'s the worker connector thread.

Reply:
Good call--I've made `onFailure` idempotent to handle this case. This approach also covers the scenario where a connector has already failed but [ConnectorContext::raiseError](https://kafka.apache.org/36/javadoc/org/apache/kafka/connect/connector/ConnectorContext.html#raiseError(java.lang.Exception)) is invoked.